File: 20-valid-parentheses/20-valid-parentheses.cpp
class Solution {
public:
    bool isValid(string s) {
        unordered_map<char, char> m = {{')', '('}, 
                                       {']', '['}, 
                                       {'}', '{'}
                                      };
        stack<char> st;
        for(char c: s){
            if(m.count(c)){
                if(st.empty() || st.top() != m[c])
                    return false;
                st.pop();
            }
            else
                st.push(c);
        }
        return st.empty();
    }
};

File: 20-valid-parentheses/20-valid-parentheses.cs
public class Solution {
    public bool IsValid(string s) {
        int n = s.Length;
        
        Stack<char> st = new Stack<char>();
        for(int i = 0; i < n; i++){
            if(s[i] == '(' || s[i] == '{' || s[i] == '[')
                st.Push(s[i]);
            else if(st.Count == 0)
                return false;
            else if(s[i] == ')'){
                if(st.Peek() != '(')
                    return false;
                else
                    st.Pop();
            }
            else if(s[i] == '}'){
                if(st.Peek() != '{')
                    return false;
                else
                    st.Pop();
            }
            else if(s[i] == ']'){
                if(st.Peek() != '[')
                    return false;
                else
                    st.Pop();
            }
        }
        
        return st.Count == 0;
    }
}